### ISSUE TITLE
ThreadScopeContext not working in DryIoc.dll 4.1.0-preview-02?

### ISSUE THREAD

**Issue by wjrogers** [dadhi/DryIoc#183]
I created a container via `new Container(scopeContext: new ThreadScopeContext())`, registered a service with `Reuse.InThread`, and tried to resolve an instance of this service. DryIoc threw this error:

> 'container.Resolve<IService>()' threw an exception of type 'DryIoc.ContainerException'
>     Data: {System.Collections.ListDictionaryInternal}
>     Error: 37
>     ErrorName: "NoCurrentScope"
>     HResult: -2146233079
>     HelpLink: null
>     InnerException: null
>     Message: "No current scope is available: probably you are registering to, or resolving from outside of the scope. \r\nCurrent resolver context is: Container with ambient ScopeContext DryIoc.ThreadScopeContext without Scope\r\n with Rules with {TrackingDisposableTransients, CaptureContainerDisposeStackTrace, AutoConcreteTypeResolution, SelectLastRegisteredFactory} and without {ThrowOnRegisteringDisposableTransient}\r\n with FactorySelector=SelectLastRegisteredFactory\r\n with Made={FactoryMethod=ConstructorWithResolvableArguments}."
>     Source: "DryIoc"
>     StackTrace: "   at DryIoc.Throw.For[T](Int32 error, Object arg0, Object arg1, Object arg2, Object arg3) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 11109\r\n   at DryIoc.Interpreter.InterpretGetScopedViaFactoryDelegate(IResolverContext r, IList`1 args, Boolean useFec) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 2878\r\n   at DryIoc.Interpreter.TryInterpret(IResolverContext r, Expression expr, Boolean useFec, Object& result) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 2559\r\n   at DryIoc.Interpreter.TryInterpret(IResolverContext r, Expression expr, Boolean useFec, Object& result) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 2731\r\n   at DryIoc.Interpreter.TryInterpretAndUnwrapContainerException(IResolverContext r, Expression expr, Boolean useFec, Object& result) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 2490\r\n   at DryIoc.Container.ResolveAndCache(Type serviceType, IfUnresolved ifUnresolved) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 310\r\n   at DryIoc.Contai
> ner.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 266\r\n   at DryIoc.Resolver.Resolve[TService](IResolver resolver, IfUnresolved ifUnresolved) in D:\\Dev\\DryIoc\\src\\DryIoc\\Container.cs:line 6195"
>     TargetSite: {T For[T](Int32, System.Object, System.Object, System.Object, System.Object)}
> 

I'm targeting `net471`. Did I miss something, or is there a DryIoc bug here?

**Comment by dadhi** [comment]
Could you provide the full sample?

**Comment by wjrogers** [comment]
I think I figured out from reading the code and some old Bitbucket issues that I misunderstood how `ThreadScopeContext` works. It _doesn't_ automatically create an ambient scope. It only tracks the current scope per thread. I was able to implement an `IScopeContext` that automatically creates a default per-thread scope by following some examples you posted in the past.

One additional surprise, though:
https://github.com/dadhi/DryIoc/blob/c363a009c1c1528626b43b157e24455c6bf2c3c4/src/DryIoc/Container.cs#L10503-L10504
I expected this to be `ScopedTo(ThreadScopeContext.ScopeContextName)`, so resolving a thread-scoped service in a child/inner scope will return the same instance as the root thread scope. Can you explain why it isn't?

**Comment by dadhi** [comment]
> I expected this to be ScopedTo(ThreadScopeContext.ScopeContextName)

Stictly speaking yes, you are right. Will fix.

**Comment by wjrogers** [comment]
Thanks! Great library. I really like how easy it is to extend with custom features.

For anyone else reading this issue: I ended up using a different solution for my use case. What I really wanted was a singleton per thread (I am using a native library that is picky about isolating components to the thread they were created on). Using `ThreadScopeContext` gets pretty messy when combined with ASP.NET Core -- the framework creates a child scope per request, which captures the ambient `IScopeContext` scope, but the thread your controller code actually runs on may be different because of async shenanigans. I chose to inject a custom factory into my API controllers so I can resolve a thread-local instance at the call site:

```csharp
namespace Example
{
    // container.Register(
    //     typeof(IThreadLocalFactory<>),
    //     typeof(ThreadLocalFactory<>),
    //     reuse: Reuse.Singleton);

    public interface IThreadLocalFactory<T> : IDisposable
    {
        T Instance { get; }
    }

    internal sealed class ThreadLocalFactory<T> : IThreadLocalFactory<T>
    {
        private readonly ThreadLocal<T> m_local;

        private int m_disposed;

        public ThreadLocalFactory(Func<T> factory)
        {
            m_local = new ThreadLocal<T>(factory, true);
        }

        public void Dispose()
        {
            if (Interlocked.CompareExchange(ref m_disposed, 1, 0) == 1)
                return;

            foreach (var instance in m_local.Values)
            {
                if (instance is IDisposable disposable)
                    disposable.Dispose();
            }

            m_local.Dispose();
        }

        public T Instance => m_local.Value;
    }
}
```

**Comment by dadhi** [comment]
Hi again, there is no bug here with `Scoped` as `InThread`, because the name `ThreadScopeContextName` is not used by DryIoc **automatically**. 

You may use it to ensure the convention in  `OpenScope(...)` and in `ScopedTo(...)`. Or may not (the default) and allow to switch the scope context without altering the reuse in registrations, e.g. in tests or depending on framework requirements, etc.

**Comment by dadhi** [comment]
I would probably mark `InThread` as obsolete in future versions in favor of `Scoped, ScopedTo` to avoid confusion.

**Comment by dadhi** [comment]
Btw, thanks for the posting of your solution. It helped me with the fix for ThreadScopeContext for #151... hopefully.

**Comment by wjrogers** [comment]
👍 sounds like this is working as you intended
